underfs/b2/src/main/java/alluxio/underfs/b2/B2InputStream.java

```diff
@@ -0,0 +1,150 @@
+   * @param offset the stream position to begin reading from
+   * @param length the length of the stream in bytes
+   */
+  public B2InputStream(String bucketName, String key, B2StorageClient b2StorageClient, long offset,
+      long length) throws ServiceException {
+    mBucketName = bucketName;
+    mKey = key;
+    mB2StorageClient = b2StorageClient;
+    mOffset = offset;
+    mLength = length;
+
+    try {
+      final B2ContentMemoryWriter sink = getSink();
+      if (mOffset > 0) {
+        final B2DownloadByNameRequest request = B2DownloadByNameRequest.builder(mBucketName, mKey)
+            .setRange(B2ByteRange.between(mOffset, mLength)).build();
+        mB2StorageClient.downloadByName(request, sink);
+      } else {
+        final B2DownloadByNameRequest request =
+            B2DownloadByNameRequest.builder(mBucketName, mKey).build();
+        mB2StorageClient.downloadByName(request, sink);
+      }
+      mBufferedInputStream = new BufferedInputStream(new ByteArrayInputStream(sink.getBytes()));
+    } catch (B2Exception e) {
+      LOG.error(
+          "Failed to open stream with remote storage. bucket name: [{}], key: [{}], offset: [{}]",
+          mBucketName, mKey, mOffset);
+      Throwables.propagateIfPossible(e, ServiceException.class);
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    mBufferedInputStream.close();
+  }
+
+  protected B2ContentMemoryWriter getSink() {
+    return B2ContentMemoryWriter.build();
+  }
+
+  @Override
+  public int read() throws IOException {
+    int ret = mBufferedInputStream.read();
+    if (ret != -1) {
+      mOffset++;
+    }
+    return ret;
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    int ret = mBufferedInputStream.read(b, off, len);
+    if (ret != -1) {
+      mOffset += ret;
+    }
+    return ret;
+  }
+
+  /**
+   * This method leverages the ability to open a stream from B2 from a given offset. When the
+   * underlying stream has fewer bytes buffered than the skip request, the stream is closed, and
+   * a new stream is opened starting at the requested offset.
+   *
+   * @param n number of bytes to skip
+   * @return the number of bytes skipped
+   */
+  @Override
+  public long skip(long n) throws IOException {
+    if (mBufferedInputStream.available() >= n) {
+      return mBufferedInputStream.skip(n);
+    }
+    // The number of bytes to skip is possibly large, open a new stream from B2.
+    mBufferedInputStream.close();
+    mOffset += n;
+    try {
+      final B2ContentMemoryWriter sink = getSink();
+      final B2DownloadByNameRequest request = B2DownloadByNameRequest.builder(mBucketName, mKey)
+          .setRange(B2ByteRange.between(mOffset, mLength)).build();
+      mB2StorageClient.downloadByName(request, sink);
+      mBufferedInputStream = new BufferedInputStream(new ByteArrayInputStream(sink.getBytes()));
+    } catch (B2Exception e) {
+      throw new IOException(e);
```

Comment:
Can we add to the message, like `new IOException(String.format("Failed to open new stream. bucket: %s key: %s offset: %d length: %d", mBucketName, mKey, mOffset, mLength), e) `